### ISSUE TITLE
Change in community detection results depending on package version

### ISSUE THREAD

**Issue by JanaMuschinski** [NetFACS/NetFACS#5]
Hello!

I wrote up an analysis involving community detection in version 0.3.1 of the NetFACS package and got consistent results (described below). When I updated the package to 0.5.0, the results are longer consistent (between versions or within version 0.5.0 when running the analysis 10-20 times in sequence). 

min_DF is just a presence/absence matrix with 341 rows and 25 elements. There are 5 rows where no element is selected, which I think may be where the inconsistency is coming in. In the 0.3.1 version these were not dropped from analysis, but whenever I run the analyses in 0.5.0 is gives the warning:

_Removing 5 rows with 0 active elements from data._

### The code:

```
min.all.greet<- netfacs(
  data = min_DF,
  condition = NULL,
  ran.trials = 1000,
  combination.size = 3,
  use_parallel = TRUE
)

min.all.net <- netfacs.network(netfacs.data = min.all.greet,
                               min.count = 2,
                               link = "unweighted")


complot(min.all.net, hide.unconnected = TRUE, title = "Collapsed ethogram")

```
Note - I slightly changed the complot function so that any element not part of a cluster would be grey 

### Results:

I ran this 20 times in version 0.3.1 and got the following results:

17x out of 20;
![image](https://user-images.githubusercontent.com/63369324/211169036-d9d25b81-cc4f-4ec9-956e-a3dda49b7ece.png)

3x out of 20:
![image](https://user-images.githubusercontent.com/63369324/211169045-e7a4e1ed-fba2-4554-bfeb-0e0012411ae1.png)

I then ran it 10 times in version 0.5.0 and got an even spread across the following results:

![image](https://user-images.githubusercontent.com/63369324/211169063-3c128d15-98db-4693-aca8-8de685bd23f5.png)
![image](https://user-images.githubusercontent.com/63369324/211169067-fbcce0bd-f40d-4353-bca4-1713ddc9a668.png)
![image](https://user-images.githubusercontent.com/63369324/211169070-8cbf9336-6121-492a-8ced-bab2391b7402.png)
![image](https://user-images.githubusercontent.com/63369324/211169073-eb4b2bd8-11a2-4aa8-a7f8-b6d14d6d0414.png)

The spread was nearly perfectly even (one came up 3x, the others all 2)

### Attempted solutions:

This issue is not resolved through any of the following:

1. Increasing the number of trials run from 1000 to 3000
2. Adding a dummy variable that is set to 1 for all of the 341 observations so none get thrown out
3. Adding a dummy variable that is set to 1 only for those that have no other elements selected and 0 for everything else

Do you see anything in the code that might be causing these issues? I was about to submit a paper relying on 0.3.1, but want to figure out why the results aren't consistent across the two versions first.

**Comment by avrincon** [comment]
Hi Jana,

Thanks for the info. It's not immediately clear to me why the results are different across versions, but I will look into it and get back to you. 

Just to note that in version 0.3.1, `netfacs()` always removed empty rows from the data, but did so silently. In a later version I added the message so that the user is aware that that is happening. So that is not the reason for the difference in results.

**Comment by avrincon** [comment]
When I run the community detection using the same `set.seed` with the different versions of NetFACS I get the same result. Do you also get the same result when the `seed` is kept constant?

Your `complot` function is not part of NetFACS so I cannot see what it is doing. 


### Here's the code:
```
# make it reproducible
seed <- 179
# install older version of package
remotes::install_version('NetFACS', version = '0.3.1')

library(NetFACS)
# Check the version of package in use
packageVersion('NetFACS')

set.seed(seed)
min.all.greet <- netfacs(
  data = emotions_set[[1]],
  condition = NULL,
  ran.trials = 10,
  combination.size = 3
)

set.seed(seed)
min.all.net <- netfacs.network(netfacs.data = min.all.greet,
                               min.count = 2,
                               link = "unweighted")

set.seed(seed)
network.plot(
  min.all.net,
  clusters = TRUE,
  plot.bubbles = TRUE
)

# Detach the current package
detach("package:NetFACS")	


# install the newer version of package
remotes::install_version('NetFACS', version = '0.5.0')
library(NetFACS)
packageVersion('NetFACS')

set.seed(seed)
min.all.greet <- netfacs(
  data = emotions_set[[1]],
  condition = NULL,
  ran.trials = 10,
  combination.size = 3
)

set.seed(seed)
min.all.net <- netfacs_network(netfacs.data = min.all.greet,
                               min.count = 2,
                               link = "unweighted")

set.seed(seed)
network_plot(
  min.all.net,
  clusters = TRUE,
  plot.bubbles = TRUE
)
```

**Comment by JanaMuschinski** [comment]
Thanks so much for the response. Good news - everything works with both versions with my data using the slimmed down version of the code (even if I use my custom plotting function). Bad news - It still has the same problems when I run it with the full analysis code for the paper. Meaning with v0.3.1 I get the "correct" results, with v0.5.0 I get different results, even though the same seed is set. I think it may be due to something in a dependent package getting masked by something else I load after loading NetFACs or something along these lines. But at this point I think it least is clear it's not a NetFACs issue, but something else going on. I'll let you know if I figure out what it is! I may just have to split things down more into separate files or unload packages between sections to make sure I don't run into issues with this.

**Comment by NetFACS** [comment]
I had some functions crash recently because they were trying to use a map() function that wasn't purrr, so I wonder if we should specify packages throughout.
________________________________

**Comment by avrincon** [comment]
If you manage to give me a reproducible example of the error, perhaps using the `emotions_set` data, I could try to work on a solution.

**Comment by JanaMuschinski** [comment]
OK, here we go. It took ages to get this to break on someone else's computer, but I finally managed it hahaha. I tested this with my own dataset on both my and a friends computer and it caused issues both times. I tried it with the emotions dataset on just mine. Below is the code from the version using the emotions dataset. 

I originally thought it was because dplyr was loaded and masking things, but it appears to do this even if you comment out library(dplyr)

```
# make it reproducible
seed <- 179


#install.packages("pacman")
pacman::p_loaded()   # check if there are currently any packages loaded

# install older version of package
remotes::install_version('NetFACS', version = '0.5.0')


library(dplyr)
library(NetFACS)


# Check the version of package in use of NetFACS
packageVersion('NetFACS')


# Reproducible example #######


set.seed(seed)


min.all.greet <- netfacs(
  data = emotions_set[[1]],
  condition = NULL,
  ran.trials = 10,
  combination.size = 3
)

set.seed(seed)
min.all.net <- netfacs_network(netfacs.data = min.all.greet,
                               min.count = 2,
                               link = "unweighted")

set.seed(seed)
network_plot(
  min.all.net,
  clusters = TRUE,
  plot.bubbles = TRUE
)


# Detach the current package
detach("package:NetFACS")	

rm(min.all.greet, min.all.net)
# install the newer version of package
remotes::install_version('NetFACS', version = '0.3.1')
library(NetFACS)
packageVersion('NetFACS')

set.seed(seed)
min.all.greet <- netfacs(
  data = emotions_set[[1]],
  condition = NULL,
  ran.trials = 10,
  combination.size = 3
)

# NOTE here it requires me to change the netfacs_network function to netfacs.network (and network.plot) BUT
# it only does this if the order is 0.5.0 version first, then 0.3.1 after. V0.3.1 seems to run with netfacs_network if you 
# load it first right after restarting a session (at least on my computer). 


set.seed(seed)
min.all.net <- netfacs.network(netfacs.data = min.all.greet,
                               min.count = 2,
                               link = "unweighted")



set.seed(seed)
network.plot(
  min.all.net,
  clusters = TRUE,
  plot.bubbles = TRUE
)
```


My session info in case it's useful:

![image](https://user-images.githubusercontent.com/63369324/213425842-a6759466-4b17-4d80-9190-735fbdba4ed5.png)

**Comment by AlexMielke1988** [comment]
Okay, so the difference does not arise in netfacs() itself but in the networking functions. one difference between the netfacs_network() vs netfacs.network() functions is that in the old version (netfacs.network()), more combinations are filtered out early on. This seems to be the case due to rounding - in the previous NetFACS version (0.3.1), things were rounded to two decimals, so many combinations that are 0.001 (because they occur once) are filtered, especially if no minimal values are set (e.g., min.count). The two combinations that appear in 0.5.0 but not in 0.3.1 (4_9 and 4_24, at least for me) are significant in both, but only at level 0.05 in the old version but at 0.01 in the new version. This could again quite easily be a rounding error. I think the results from 0.5.0 are more accurate under those circumstances - the question is whether a significance level of 0.01 or 0.05 is a better fit in Jana's case given the small sample.

**Comment by JanaMuschinski** [comment]
Alright, I'm not 100% sure I understand what's going on - so there is a rounding issue somewhere that is putting things above/below certain thresholds. But even when I put those thresholds quite high (so reducing number of decimal points involved, hopefully taking out the potential for a rounding error to affect it), I'm not getting the same results between the two versions. 

Example here - same code as above, but the network function is as follows. I increased min.count and min.prob. I also tried both raising the significance threshold to 0.05 and decreasing it to 0.001 for both (not shown in code, but shown in figures below). I don't get the same results between the two versions in either scenario. Results are consistent within each version regardless of whether I use a significance cutoff of 0.001 or 0.05. They do change if I change min count and min prob, but still couldn't get them to match up after trying several different combinations (mainly increasing both min count and min prob). 

So is it that the rounding error was a problem in 0.3.1 and is resolved in 0.5.0?

```
# 0.5.0 version
min.all.net <- netfacs_network(netfacs.data = min.all.greet,
                               significance = 0.05,
                               min.count = 5,
                               link = "unweighted",
                               min.prob = 0.1)

# 0.3.1 version
min.all.net <- netfacs.network(netfacs.data = min.all.greet,
                               min.count = 5,
                               link = "unweighted",
                               min.prob = 0.1)


```

With the emotions dataset I get the following:

From 0.5.0 with 0.05 significance level
![image](https://user-images.githubusercontent.com/63369324/214418025-8dbbc584-20af-4a83-9a88-15abdd6253f3.png)

From 0.5.0 with 0.001 significance level 
![image](https://user-images.githubusercontent.com/63369324/214418981-e2425e31-33f5-43f2-8088-695a8931d618.png)


From 0.3.1 with 0.05 significance level
![image](https://user-images.githubusercontent.com/63369324/214418399-bc751f17-8d47-49d8-b613-29d561efcc50.png)

From 0.3.1 with 0.001 significance level
![image](https://user-images.githubusercontent.com/63369324/214418587-56b53565-21be-4990-898a-078b39ec7c53.png)

**Comment by JanaMuschinski** [comment]
Alex explained issue was a rounding error in 0.3.1 causing things to artificially fall below the 0.01 significance cutoff. This is not the case in 0.5.0 and so those communities don't show up (since they wouldn't meet the significance threshold).
